Ilariae/settlement tutorial (#350)
wormhole-foundation/wormhole-docs: tutorials/by-product/settlements/settlement-routes.md

---
title: Settlement with Mayan Swift
description: Learn how to integrate Wormhole Settlement Routes using the SDK to simplify cross-chain swaps, manage fees, and execute seamless transactions.
categories: Settlement, Transfer
---

# Integrate Mayan Swift Routes using the SDK

:simple-github: [Source code on GitHub](https://github.com/wormhole-foundation/demo-mayanswift){target=\_blank}

## Introduction

This tutorial explains how to integrate the Wormhole Mayan Swift Route from the Wormhole SDK into your application. This Route abstracts the complexity of cross-chain token swaps, handling route discovery, fee estimation, and transaction construction.

[Mayan Swift](https://mayan.finance/){target=\_blank} s a cross-chain swap auction protocol that simplifies bridging and swapping tokens across blockchains to offer the best swap rates using an auction mechanism.

!!!note

>     Mayan Swift currently supports **MainNet** only. Attempting to run the demo on testnet will result in failure.

Reviewer: ```suggestion
    Mayan Swift currently supports **mainnet** only. Attempting to run the demo on testnet will result in failure.
```

## Prerequisites

Before you begin, ensure you have the following:

- [Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm){target=\_blank} installed on your machine
- A wallet with a private key, funded with native tokens on mainnet for gas fees

## Project Setup

> To get started, you’ll first clone the repository, install dependencies, and configure environment variables. Once set up, you can initiate a cross-chain token swap using the Mayan Swift Route.

Reviewer: ```suggestion
To get started, you’ll first clone the repository, install dependencies, and configure environment variables. Once set up, you can initiate a multichain token swap using the Mayan Swift Route.
```

1. **Initialize the project** - start by creating a new directory for your project and initializing it with `npm`, which will create the `package.json` file for your project

    ```bash
    mkdir demo-mayanswift
    cd demo-mayanswift
    npm init -y
    ```

> 1. **Install dependencies**

Reviewer: No need for `npm install` if we don't have any dependencies in the `package.json` yet

    ```bash
    npm install
    ```

3. **Set up environment variables** - to securely store your private key, create a `.env` file in the root of your project

    ```bash
    touch .env
    ```

    Inside the `.env` file, add your private keys.

    ```env
    ETH_PRIVATE_KEY="INSERT_YOUR_PRIVATE_KEY"
    SOL_PRIVATE_KEY="INSERT_YOUR_PRIVATE_KEY"
    ```

    !!!note

        Ensure your private key contains native tokens for gas on both the source and destination chains. For Sui, you must provide a mnemonic instead of a private key.

> 4. **Create a `helpers.ts` file** - retrieves the correct signer (for Solana or an EVM chain) from environment variables based on the provided chain context, and returns that signer and its associated chain address

Reviewer: This is not only for Solana and EVM chains. Aptos and Sui can also be included here. If you only mention Solana and EVM, say that it's because you will be working with only those 2.

        ```bash
        mkdir -p src
        touch src/helpers.ts
        ```

    2. Open the `helpers.ts` file and add the following code

        ```typescript
        --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/helper.ts"
        ```

        - `getEnv` - reads environment variables from your local `.env` file
        - `getSigner` - main utility function to get a correct signer based on the chain you’re targeting. The signer is responsible for signing transactions and interacting with the blockchain. It securely uses the private key stored in your `.env` file

## Create Swap Script

This section shows how to initiate and complete a cross-chain token swap using the Mayan Swift route from the Wormhole SDK. You’ll fetch token details, calculate a quote, and sign the transaction—all with a few lines of code.

### Overview and Setup

1. **Create the `swap.ts` file** - from your project’s root directory, create a new file in the src folder. This file will handle the entire end-to-end swap process

    ```bash
    touch src/swap.ts
    ```

2. **Add your import statements** - in `swap.ts`, import the necessary libraries:
 
    ```typescript

>     --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:1:14"

Reviewer: You need to add a step in the project setup to install all of these dependencies:
`npm install @mayanfinance/wormhole-sdk-route @wormhole-foundation/sdk-solana ...`

    ```

### Implement the Swap Logic

1. **Define the Wormhole environment** - initialize your Wormhole client for Mainnet and load the desired platforms (EVM, Solana, etc.). Also define your source and destination chains. 

    !!!note
        We’ll start by creating an immediately invoked asynchronous function ((async function() { ... })()) so we can use the await keyword inside our script. This structure makes the code self-contained and allows you to call asynchronous SDK methods without extra boilerplate.

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:16:25"

    // ... in later steps, you’ll add more code here ...

    })();
    ```

> 2. **Inspect aupported tokens** - checking the tokens that each route can handle helps confirm your chain selections are valid. This step is optional but is often useful for debugging or logging

Reviewer: ```suggestion
2. **Inspect supported tokens** - checking the tokens that each route can handle helps confirm your chain selections are valid. This step is optional but is often useful for debugging or logging
```

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:31:40"
    ```

3. **Retrieve signers** - fetch the appropriate signers for both the source and destination chains using the `getSigner`helper. You’ll need these to sign transactions on each chain

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:42:44"
    ```

> 4. **Create a transfer request** - construct a `RouteTransferRequest` object that tells the Wormhole SDK what you want to transfer (source token) and what you want to receive (destination token). The tr object will be reused in subsequent steps to identify which tokens you’re transferring and to which networks

Reviewer: It is unclear what the user needs to input in the parameters (source and destination). Explicitly say that it needs to be the token address if that's the case. 

Also, I would remove the comments from the snippets and add them in the steps if necessary.

Author: it doesn't need to input anything here, source and destination are the tokenids defined at the beginning 

```
const source = Wormhole.tokenId(sendChain.chain, "native");
 const destination = Wormhole.tokenId(destChain.chain, "native");
```

> 4. **Create a transfer request** - construct a `RouteTransferRequest` object that tells the Wormhole SDK what you want to transfer (source token) and what you want to receive (destination token). The tr object will be reused in subsequent steps to identify which tokens you’re transferring and to which networks

Reviewer: ```suggestion
4. **Create a transfer request** - construct a `RouteTransferRequest` object that tells the Wormhole SDK what you want to transfer (source token) and what you want to receive (destination token). The `tr` object will be reused in subsequent steps to identify which tokens you’re transferring and to which networks
```

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:46:51"
    ```

5. **Find and select a route** - use the resolver to identify all routes capable of bridging your chosen tokens. Then pick one—commonly the “best” route

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:53:55"
    ```

>     The Wormhole `resolver.findRoutes()` function returns an array of potential routes, each with different properties. Mayan Swift typically appears if your chosen tokens and chains are supported

Reviewer: Why do you choose the first element of the array `[0]`? Is that always going to be the Mayan Swift?

Author: resolver.findRoutes(tr) asks every registered route implementation if they can fulfill the request
It returns all routes that say yes, then sorts that list to the most attractive offer. So foundRoutes[0] is the highest‑ranked candidate at that moment - I'll add explanation

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:78:85"
    ```

    The `initiate()` call will usually return a transaction receipt or identifier. You can log it or store it for reference

10. **Complete the transfer (wait for finality)** - finally, wait for the transaction to finalize on the source chain and complete the process on the destination chain. The `checkAndCompleteTransfer()` method polls the network until the cross-chain transaction is confirmed, delivering tokens to the destination address

    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts:87"
    ```

??? code "Complete script"
    ```typescript
    --8<-- "code/tutorials/by-product/settlement/demo-mayanswift/swap.ts"
    ```

## Configuration

You can customize the following options within the scripts:

- **Source and destination chains** - modify `sendChain` and `destChain` in `swap.ts`
- **Amount and transfer settings** - adjust `amount` to suit your needs

## Execute the Swap

To initiate a token transfer across chains, using the Mayan Swift Route run:

```bash

> npm run swap

Reviewer: This doesn't work because you haven't defined a `script` for this in the `package.json`
